packages/phone-number-privacy/signer/src/pnp/endpoints/sign/action.ts

```diff
@@ -0,0 +1,109 @@
+import { SignMessageRequest } from '@celo/identity/lib/odis/query'
+import { WarningMessage } from '@celo/phone-number-privacy-common'
+import { Knex } from 'knex'
+import { Action, Session } from '../../../common/action'
+import { computeBlindedSignature } from '../../../common/bls/bls-cryptography-client'
+import { getRequestExists } from '../../../common/database/wrappers/request'
+import { DefaultKeyName, Key, KeyProvider } from '../../../common/key-management/key-provider-base'
+import { Counters } from '../../../common/metrics'
+import { SignerConfig } from '../../../config'
+import { PnpQuotaService } from '../../services/quota'
+import { PnpSession } from '../../session'
+import { PnpSignIO } from './io'
+import { LegacyPnpSignIO } from './io.legacy'
+
+export class PnpSignAction implements Action<SignMessageRequest> {
+  constructor(
+    readonly db: Knex,
+    readonly config: SignerConfig,
+    readonly quota: PnpQuotaService,
+    readonly keyProvider: KeyProvider,
+    readonly io: PnpSignIO | LegacyPnpSignIO
+  ) {}
+
+  public async perform(session: PnpSession<SignMessageRequest>): Promise<void> {
+    await this.db.transaction(async (trx) => {
+      let queryCount, totalQuota, blockNumber
+      if (await getRequestExists(this.db, session.request.body, session.logger, trx)) {
+        Counters.duplicateRequests.inc()
+        session.logger.debug(
+          'Request already exists in db. Will service request without charging quota.'
+        )
+        session.errors.push(WarningMessage.DUPLICATE_REQUEST_TO_GET_PARTIAL_SIG)
+      } else {
+        const quotaStatus = await this.quota.getQuotaStatus(session, trx)
+        queryCount = quotaStatus.queryCount
+        totalQuota = quotaStatus.totalQuota
+        blockNumber = quotaStatus.blockNumber
+        // In the case of a blockchain connection failure, totalQuota and/or blockNumber
+        // may be undefined.
+        // In the case of a database connection failure, queryCount
+        // may be undefined.
+        if (quotaStatus.queryCount && quotaStatus.totalQuota) {
```

Comment:
TODO double check this: probably want to check that these fields are `undefined`. I think this currently could lead to some weird behavior if queryCount or totalQuota == 0 (at least for queryCount, we may not ever update a user that hasn't already completed at least one query?)